src/hotspot/share/opto/locknode.cpp

```diff
@@ -51,19 +51,48 @@ BoxLockNode::BoxLockNode( int slot ) : Node( Compile::current()->root() ),
   _inmask.Insert(reg);
 }
 
-//-----------------------------hash--------------------------------------------
 uint BoxLockNode::hash() const {
-  if (EliminateNestedLocks)
+  if (EliminateNestedLocks) {
     return NO_HASH; // Each locked region has own BoxLock node
-  return Node::hash() + _slot + (_is_eliminated ? Compile::current()->fixed_slots() : 0);
+  }
+  return Node::hash() + _slot + (is_eliminated() ? Compile::current()->fixed_slots() : 0);
 }
 
-//------------------------------cmp--------------------------------------------
 bool BoxLockNode::cmp( const Node &n ) const {
-  if (EliminateNestedLocks)
+  if (EliminateNestedLocks) {
     return (&n == this); // Always fail except on self
+  }
   const BoxLockNode &bn = (const BoxLockNode &)n;
-  return bn._slot == _slot && bn._is_eliminated == _is_eliminated;
+  return (bn._slot == _slot) && (bn.is_eliminated() == is_eliminated());
+}
+
+Node* BoxLockNode::Identity(PhaseGVN* phase) {
+  if (!EliminateNestedLocks && !this->is_eliminated()) {
+    Node* n = phase->hash_find(this);
+    if (n == nullptr || n == this) {
+      return this;
+    }
+    BoxLockNode* old_box = n->as_BoxLock();
+    // Set corresponding status (_kind) when commoning BoxLock nodes.
+    if (this->_kind != old_box->_kind) {
+      if (this->is_unbalanced()) {
+        old_box->set_unbalanced();
+      }
+      if (!old_box->is_unbalanced()) {
+        // Only Regular or Coarsened status should be here:
+        // Nested and Local are set only when EliminateNestedLocks is on.
+        if (old_box->is_regular()) {
+          assert(this->is_coarsened(),"unexpected kind: %s", _kind_name[(int)this->_kind]);
+          old_box->set_coarsened();
+        } else {
+          assert(this->is_regular(),"unexpected kind: %s", _kind_name[(int)this->_kind]);
+          assert(old_box->is_coarsened(),"unexpected kind: %s", _kind_name[(int)old_box->_kind]);
+        }
+      }
+    }
+    return old_box;
+  }
+  return this;
```

Comment:
For `EliminateNestedLocks` we now never have a hash, are only equal with `this`, and don't common nodes in `Ideal` either. Is that intended?

Are there any IR tests that verify that we common nodes in the cases where we expect it to common?

Reply:
> For EliminateNestedLocks we now never have a hash, are only equal with this, and don't common nodes in Ideal either. Is that intended?

It was since JDK 8. I only corrected style there: added `{}`. And replaced `_is_eliminated` with accessor call.

It is intended for nested locks elimination - to have only one Object per locking region: [JDK-7125896@locknode.cpp](https://github.com/openjdk/jdk/commit/94927c382b40c31f4525b0c71a64e34c26e03b98#diff-d38a64764056a8888cdffdfbbd8aa31cf00ddc96040bae83112521559a7fa0a6)

> Are there any IR tests that verify that we common nodes in the cases where we expect it to common?

These optimizations were implemented before we had IR tests. We have some regression tests. During my experiments I added `_kind` to `hash()` and a lot of tests in Tier1 (not just compiler) failed when ran with `-XX:-EliminateNestedLocks`:
```
#  Internal Error (/workspace/open/src/hotspot/share/opto/parse1.cpp:2030), pid=71294, tid=28931
#  assert(!nocreate) failed: Cannot build a phi for a block already parsed.
```
So we have this verification that we common these nodes.